function-images/tests/video-bench/decoder/decode.py

```diff
@@ -0,0 +1,210 @@
+            if XDTconfig is None:
+                log.fatal("Empty XDT config")
+            self.XDTconfig = XDTconfig
+
+    def Decode(self, request, context):
+        log.info("Decoder recieved a request")
+
+        s3 = None
+        videoBytes = b''
+        if self.transferType == S3:
+            log.info("Using s3, getting bucket")
+            with tracing.Span("Video fetch"):
+                s3, videoBytes = fetchFromS3(self.s3_client, request.s3key)
+            log.info("decoding frames of the s3 object")
+        elif self.transferType == INLINE:
+            log.info("Inline video decode. Decoding frames.")
+            videoBytes = request.video
+        results = self.processFrames(videoBytes)
+        return videoservice_pb2.DecodeReply(classification=results)
+
+    def processFrames(self, videoBytes):
+        frames = decode(videoBytes)
+        with tracing.Span("Recognise all frames"):  
+            all_result_futures = []
+            # send all requests
+            for i in range(int(os.getenv('DecoderFrames', int(args.frames)))):
+                if self.transferType == S3 or self.transferType == INLINE:
+                    all_result_futures.append(self.Recognise(frames[i]))
+                if self.transferType == XDT:
+                    all_result_futures.append(self.Recognise(frames[i]))
+            # concat all results
+            log.info("returning result of frame classification")
+            results = ""
+            for result in all_result_futures:
+                results = results + result + ","
+
+            return results
+    
+    def Recognise(self, frame):
+        channel = grpc.insecure_channel(args.addr)
+        stub = videoservice_pb2_grpc.ObjectRecognitionStub(channel)
+        if self.transferType == S3:
+            name = "decoder-frame-" + str(self.frameCount) + ".jpg"
+            with tracing.Span("Upload frame"):
+                s3object = self.s3_client.Object('vhive-video-bench', name)
+                self.frameCount += 1
+                log.info("uploading frame %d to s3" % self.frameCount)
+                s3object.put(Body=frame)
+            log.info("calling recog with s3 key")
+            response_future = stub.Recognise.future(videoservice_pb2.RecogniseRequest(s3key=name))
+            result = response_future.result().classification
+        elif self.transferType == INLINE:
+            response_future = stub.Recognise.future(videoservice_pb2.RecogniseRequest(frame=frame))
+            result = response_future.result().classification
+        elif self.transferType == XDT:
+            xdtPayload = XDTutil.Payload(FunctionName="HelloXDT", Data=frame)
+            if not args.dockerCompose:
+                log.info("replacing SQP hostname")
+                self.XDTconfig["SQPServerHostname"] = get_self_ip()
+            response_future, ok = XDTsrc.InvokeWithXDT(args.addr, xdtPayload, self.XDTconfig)
+            result = response_future.decode()
```

Comment:
could you please clarify this?

Reply:
InvokeWithXDT returns bytes. `.decode` is used to convert bytes to string.

Reply:
leave a comment please